ding/framework/supervisor.py

```diff
@@ -0,0 +1,344 @@
+            args=(self._proc_id, self._init, self._args, self._send_queue, self._recv_queue, self._shared_object),
+            name="supervisor_child_{}_{}".format(self._proc_id, time.time()),
+            daemon=True
+        )
+        proc.start()
+        self._proc = proc
+
+    def shutdown(self, timeout: Optional[float] = None):
+        if self._proc:
+            self._send_queue.put(SendPayload(proc_id=self._proc_id, method=ReserveMethod.SHUTDOWN))
+            self._proc.terminate()
+            self._proc.join(timeout=timeout)
+            if hasattr(self._proc, "close"):  # Compatible with 3.6
+                self._proc.close()
+            self._proc = None
+            self._send_queue.close()
+            self._send_queue.join_thread()
+            self._send_queue = None
+
+    def send(self, payload: SendPayload):
+        self._send_queue.put(payload)
+
+
+class ChildThread(Child):
+
+    def __init__(self, proc_id: int, init: Callable, *args, shared_object: Optional[SharedObject] = None) -> None:
+        super().__init__(proc_id, init, *args, shared_object=shared_object)
+        self._thread = None
+
+    def start(self, recv_queue: queue.Queue):
+        self._recv_queue = recv_queue
+        self._send_queue = queue.Queue()
+        thread = threading.Thread(
+            target=self._target,
+            args=(self._proc_id, self._init, self._args, self._send_queue, self._recv_queue),
+            name="supervisor_child_{}_{}".format(self._proc_id, time.time()),
+            daemon=True
+        )
+        thread.start()
+        self._thread = thread
+
+    def shutdown(self, timeout: Optional[float] = None):
+        if self._thread:
+            self._send_queue.put(SendPayload(proc_id=self._proc_id, method=ReserveMethod.SHUTDOWN))
+            self._thread.join(timeout=timeout)
+            self._thread = None
+            self._send_queue = None
+
+    def send(self, payload: SendPayload):
+        self._send_queue.put(payload)
+
+
+class Supervisor:
+
+    TYPE_MAPPING = {ChildType.PROCESS: ChildProcess, ChildType.THREAD: ChildThread}
+
+    QUEUE_MAPPING = {ChildType.PROCESS: mp.Queue, ChildType.THREAD: queue.Queue}
+
+    def __init__(self, type_: ChildType) -> None:
+        self._children: List[Child] = []
+        self._type = type_
+        self._child_class = self.TYPE_MAPPING[self._type]
+        self._running = False
+        self.__queue = None
+
+    def register(self, init: Callable, *args, shared_object: Optional[SharedObject] = None) -> None:
+        proc_id = len(self._children)
+        self._children.append(self._child_class(proc_id, init, *args, shared_object=shared_object))
+
+    @property
+    def _recv_queue(self) -> Union[queue.Queue, mp.Queue]:
+        if not self.__queue:
+            self.__queue = self.QUEUE_MAPPING[self._type]()
+        return self.__queue
+
+    @_recv_queue.setter
+    def _recv_queue(self, value):
+        self.__queue = value
+
+    def start_link(self) -> None:
```

Comment:
also name as `start`?

Reply:
The idea of supervisor is borrowed from [gen_server](https://www.erlang.org/doc/man/gen_server.html#start_link-3), so just keep the original method name.